src/main/java/com/netflix/conductor/kafka/eventqueue/KafkaObservableQueue.java

```diff
@@ -0,0 +1,408 @@
+        for (Message message : messages) {
+            String[] idParts = message.getId().split(":");
+            int partitionNumber = Integer.parseInt(idParts[2]);
+            for (String topic : this.topics) {
+                for (PartitionInfo partition : consumer.partitionsFor(topic)) {
+                    if (partitionNumber == partition.partition()) {
+                        Map<TopicPartition, OffsetAndMetadata> currentOffsets = new HashMap<>();
+                        currentOffsets.put(new TopicPartition(idParts[1], partitionNumber),
+                                new OffsetAndMetadata(Integer.parseInt(idParts[3]) + 1, "no metadata"));
+                        try {
+                            consumer.commitSync(currentOffsets);
+                        } catch (KafkaException ke) {
+                            messageIds.add(message.getId());
+                            logger.error("kafka consumer selective commit failed.", ke);
+                        }
+                    }
+                }
+            }
+        }
+        return messageIds;
+    }
+
+    public void setUnackTimeout(Message message, long unackTimeout) {
+    }
+
+    @Override
+    public void publish(List<Message> messages) {
+        publishMessages(messages);
+    }
+
+    @Override
+    public long size() {
+        return 0;
+    }
+
+    @Override
+    public String getType() {
+        return QUEUE_TYPE;
+    }
+
+    @Override
+    public String getName() {
+        return queueURI;
+    }
+
+    @Override
+    public String getURI() {
+        return queueURI;
+    }
+
+    /**
+     * Polls the topics and retrieve the messages for all consumers of the topic.
+     *
+     * @return List of messages
+     */
+    List<Message> receiveMessages() {
+        return receiveMessages(consumer);
+    }
+
+    /**
+     * Polls the topics and retrieve the messages for a consumer.
+     *
+     * @return List of messages
+     */
+    List<Message> receiveMessages(KafkaConsumer<String, String> consumer) {
+        List<Message> messages = new ArrayList<>();
+        try {
+
+            ConsumerRecords<String, String> records = consumer.poll(Duration.ofMillis(pollTimeoutInMs));
+
+            if (records.count() == 0) {
+                return messages;
+            }
+
+            logger.info("polled {} messages from kafka topic.", records.count());
+            records.forEach(record -> {
+                Map<String, String> headers = new HashMap<>();
+                record.headers().forEach(header -> headers.put(header.key(),
+                        new String(header.value(), StandardCharsets.UTF_8)));
+                if (this.filteringHeader != null && !this.filteringValue.equals(headers.get(this.filteringHeader))
+                        || this.filterByPartitionKey && !this.filteringValue.equals(record.key())) {
+                    addAsyncCommitOffset(record.topic(), record.partition(), record.offset());
```

Comment:
I didn't get why only message which filtered out are saved to threal local map? For my understanding all message should be proceed via this mechamism.